### DISCUSSION TITLE
Can ROS2 image transfer possible using zero copy communication

### DISCUSSION THREAD

**Discussion by vishnuv-dev** [eProsima/Fast-DDS#5314] [General]
I am referring the following document for zero copy communication [https://fast-dds.docs.eprosima.com/en/latest/fastdds/use_cases/zero_copy/zero_copy.html](url) to work on ROS2 image transfer. In this document, it is mentioned that zero copy communication has some constrains like "only plain types are supported". Will this affect the image transfer between nodes using zero copy communication.?

**Comment by MiguelCompany** [comment]
The standard message types (like [this one](https://github.com/ros2/common_interfaces/blob/jazzy/sensor_msgs/msg/Image.msg))  contain a string in the header that makes them unusable with zero copy.

You might be interested in [this project](https://github.com/ZhenshengLee/ros2_shm_msgs) which contains an image processing pipeline fully compatible with zero copy.

**Reply by narol1024** [reply]
@MiguelCompany Thank you, this looks good, but custom messages might not be compatible with other modules. For example, Foxglove uses string type `encoding` when parsing messages, which could cause incompatibility. I’m not sure how to make them compatible.

**Reply by MiguelCompany** [reply]
Apart from my previous suggestion, if you don't care building ROS 2 from sources, you could take a look at the PoC mentioned [here](https://github.com/ros2/rclcpp/issues/2201#issuecomment-1640125818)
